File: frontend/src/utils/roles.js
export const ROLES = {
  USER: 'USER',
  ADMIN: 'ADMIN',
  SUPER_ADMIN: 'SUPER_ADMIN',
};

export const normalizeRole = (role) => {
  if (!role) return ROLES.USER;
  if (role === 'SUPER_ADMIN' || role === 'SUPERADMIN') return ROLES.SUPER_ADMIN;
  if (role === 'ADMIN') return ROLES.ADMIN;
  return ROLES.USER;
};

export const getRoleFlags = (role) => {
  const r = normalizeRole(role);
  const isSuperAdmin = r === ROLES.SUPER_ADMIN;
  const isAdmin = r === ROLES.ADMIN;
  const isUser = r === ROLES.USER;
  const isStaff = isAdmin || isSuperAdmin;

  return {
    role: r,
    isSuperAdmin,
    isAdmin,
    isUser,
    isStaff,
    canLeaveChat: isSuperAdmin,
    canAccessAdminPortal: isSuperAdmin,
    canAccessChatNav: isSuperAdmin,
    canAccessTimetables: isSuperAdmin,
    canBypassChatPasscode: isSuperAdmin,
    canUseTaskManager: isStaff,
  };
};

const TIMETABLE_PAGES = new Set([
  'timetables',
  'class_timetable',
  'cat_timetable',
  'national_exam',
  'dining_timetable',
]);

/** Whether a sidebar page id is allowed for this role */
export const canAccessPage = (role, pageId) => {
  const flags = getRoleFlags(role);

  if (flags.isUser) {
    return pageId === 'chat';
  }

  if (flags.isAdmin) {
    if (pageId === 'chat' || pageId === 'admin') return false;
    if (TIMETABLE_PAGES.has(pageId)) return false;
    return true;
  }

  if (flags.isSuperAdmin) {
    return true;
  }

  return false;
};

/** USER with chat goes straight to chat; staff land on dashboard */
export const getDefaultPageForUser = (user) => {
  const flags = getRoleFlags(user?.role);
  if (flags.isUser) {
    return user?.chatEnabled ? 'chat' : 'chat';
  }
  return 'dashboard';
};

export const isChatOnlyUser = (user) =>
  getRoleFlags(user?.role).isUser && !!user?.chatEnabled;

/** Direct messages / chat picker — USER (chat enabled) and SUPER_ADMIN only; never ADMIN */
export const canAppearInChatList = (role, chatEnabled = false) => {
  const r = normalizeRole(role);
  if (r === ROLES.ADMIN) return false;
  if (r === ROLES.SUPER_ADMIN) return true;
  if (r === ROLES.USER) return !!chatEnabled;
  return false;
};

File: backend/src/main/java/com/chat/app/security/Roles.java
package com.chat.app.security;

/**
 * Application roles:
 * USER — chat only (when chatEnabled)
 * ADMIN — task manager sidebar (no chat, admin portal, timetables)
 * SUPER_ADMIN — full access
 */
public final class Roles {

    public static final String USER = "USER";
    public static final String ADMIN = "ADMIN";
    public static final String SUPER_ADMIN = "SUPER_ADMIN";

    private Roles() {
    }

    public static String normalize(String role) {
        if (role == null || role.isBlank()) {
            return USER;
        }
        role = role.trim();
        if (SUPER_ADMIN.equalsIgnoreCase(role) || "SUPERADMIN".equalsIgnoreCase(role)) {
            return SUPER_ADMIN;
        }
        if (ADMIN.equalsIgnoreCase(role)) {
            return ADMIN;
        }
        return USER;
    }

    public static boolean isSuperAdmin(String role) {
        return SUPER_ADMIN.equals(normalize(role));
    }

    public static boolean isAdmin(String role) {
        return ADMIN.equals(normalize(role));
    }

    public static boolean isUser(String role) {
        return USER.equals(normalize(role));
    }

    public static boolean isStaff(String role) {
        return isAdmin(role) || isSuperAdmin(role);
    }

    public static boolean canUseTaskManager(String role) {
        return isStaff(role);
    }

    public static boolean canUseChat(String role, boolean chatEnabled) {
        if (isSuperAdmin(role)) {
            return true;
        }
        if (isAdmin(role)) {
            return false;
        }
        return isUser(role) && chatEnabled;
    }
}
